rs/consensus/src/consensus/block_maker.rs

```diff
@@ -514,17 +516,50 @@ pub(crate) fn already_proposed(pool: &PoolReader<'_>, h: Height, this_node: Node
         .any(|p| p.signature.signer == this_node)
 }
 
+const DYNAMIC_DELAY_LOOK_BACK_DISTANCE: Height = Height::new(29);
+const DYNAMIC_DELAY_MAX_NON_RANK_0_BLOCKS: usize = 5;
+const DYNAMIC_DELAY_EXTRA_DURATION: Duration = Duration::from_secs(10);
+
+fn count_non_rank_0_blocks(pool: &PoolReader, block: Block) -> usize {
+    let max_height = block.height();
+    let min_height = if max_height > DYNAMIC_DELAY_LOOK_BACK_DISTANCE {
+        max_height - DYNAMIC_DELAY_LOOK_BACK_DISTANCE
+    } else {
+        Height::new(0)
+    };
+
+    pool.get_range(block, min_height, max_height)
+        .filter(|block| block.rank > Rank(0))
+        .count()
+}
+
 /// Calculate the required delay for block making based on the block maker's
-/// rank.
+/// rank and the number of non-0-rank blocks in its ancestry.
 pub(super) fn get_block_maker_delay(
     log: &ReplicaLogger,
     registry_client: &dyn RegistryClient,
     subnet_id: SubnetId,
+    pool: &PoolReader<'_>,
+    parent: Block,
     registry_version: RegistryVersion,
     rank: Rank,
+    metrics: Option<&BlockMakerMetrics>,
 ) -> Option<Duration> {
-    get_notarization_delay_settings(log, registry_client, subnet_id, registry_version)
-        .map(|settings| settings.unit_delay * rank.0 as u32)
+    let settings =
+        get_notarization_delay_settings(log, registry_client, subnet_id, registry_version)?;
+    let unit_delay = settings.unit_delay;
+
+    let dynamic_delay =
+        if count_non_rank_0_blocks(pool, parent) > DYNAMIC_DELAY_MAX_NON_RANK_0_BLOCKS {
+            if let Some(metrics) = metrics {
+                metrics.dynamic_delay_triggered.inc();
+            }
+            unit_delay + DYNAMIC_DELAY_EXTRA_DURATION
+        } else {
+            unit_delay
+        };
+
+    Some(dynamic_delay * rank.0 as u32)
```

Comment:
Do we really want to add 10s for each rank? Wouldn't it be enought to return `unit_delay * rank.0 as u32  + DYNAMIC_DELAY_EXTRA_DURATION`?

Reply:
hmm, you're right. I think we did agree to do `unit_delay * rank.0 as u32  + DYNAMIC_DELAY_EXTRA_DURATION` (at least this is what I put in the design doc, I somehow misread it).

I'll fix that

Reply:
Fixed. Thanks for catching it :)